protocols.go

```diff
@@ -40,156 +23,126 @@ const (
-			return fmt.Errorf("protocol by the name %q already exists", p.Name)
-		}
+	protoUTP = Protocol{
+		Name:  "utp",
+		Code:  P_UTP,
+		VCode: CodeToVarint(P_UTP),
 	}
-
-	Protocols = append(Protocols, p)
-	ProtocolsByName[p.Name] = p
-	return nil
-}
-
-// ProtocolWithName returns the Protocol description with given string name.
-func ProtocolWithName(s string) Protocol {
-	return ProtocolsByName[s]
-}
-
-// ProtocolWithCode returns the Protocol description with given protocol code.
-func ProtocolWithCode(c int) Protocol {
-	for _, p := range Protocols {
-		if p.Code == c {
-			return p
-		}
+	protoUDT = Protocol{
+		Name:  "udt",
+		Code:  P_UDT,
+		VCode: CodeToVarint(P_UDT),
 	}
-	return Protocol{}
-}
-
-// ProtocolsWithString returns a slice of protocols matching given string.
-func ProtocolsWithString(s string) ([]Protocol, error) {
-	s = strings.Trim(s, "/")
-	sp := strings.Split(s, "/")
-	if len(sp) == 0 {
-		return nil, nil
+	protoQUIC = Protocol{
+		Name:  "quic",
+		Code:  P_QUIC,
+		VCode: CodeToVarint(P_QUIC),
 	}
-
-	t := make([]Protocol, len(sp))
-	for i, name := range sp {
-		p := ProtocolWithName(name)
-		if p.Code == 0 {
-			return nil, fmt.Errorf("no protocol with name: %s", name)
-		}
-		t[i] = p
+	protoHTTP = Protocol{
+		Name:  "http",
+		Code:  P_HTTP,
+		VCode: CodeToVarint(P_HTTP),
 	}
-	return t, nil
-}
-
-// CodeToVarint converts an integer to a varint-encoded []byte
-func CodeToVarint(num int) []byte {
-	buf := make([]byte, bits.Len(uint(num))/7+1)
-	n := binary.PutUvarint(buf, uint64(num))
-	return buf[:n]
-}
-
-// VarintToCode converts a varint-encoded []byte to an integer protocol code
-func VarintToCode(buf []byte) int {
-	num, _, err := ReadVarintCode(buf)
-	if err != nil {
-		panic(err)
+	protoHTTPS = Protocol{
+		Name:  "https",
+		Code:  P_HTTPS,
+		VCode: CodeToVarint(P_HTTPS),
 	}
-	return num
-}
+	protoP2P = Protocol{
+		Name:       "ipfs",
+		Code:       P_P2P,
+		VCode:      CodeToVarint(P_P2P),
+		Size:       LengthPrefixedVarSize,
+		Transcoder: TranscoderP2P,
+	}
+	protoUNIX = Protocol{
+		Name:       "unix",
+		Code:       P_UNIX,
+		VCode:      CodeToVarint(P_UNIX),
+		Size:       LengthPrefixedVarSize,
+		Path:       true,
+		Transcoder: TranscoderUnix,
+	}
+)
 
-// ReadVarintCode reads a varint code from the beginning of buf.
-// returns the code, and the number of bytes read.
-func ReadVarintCode(buf []byte) (int, int, error) {
-	num, n := binary.Uvarint(buf)
-	if n < 0 {
-		return 0, 0, fmt.Errorf("varints larger than uint64 not yet supported")
+func init() {
+	for _, p := range []Protocol{
+		protoIP4,
+		protoTCP,
+		protoUDP,
+		protoDCCP,
+		protoIP6,
+		protoSCTP,
+		protoONION,
+		protoUTP,
+		protoUDT,
+		protoQUIC,
+		protoHTTP,
+		protoHTTPS,
+		protoP2P,
+		protoUNIX,
+	} {
+		if err := AddProtocol(p); err != nil {
+			panic(err)
+		}
 	}
-	return int(num), n, nil
+
+	// explicitly set both of these
+	protocolsByName["p2p"] = protoP2P
+	protocolsByName["ipfs"] = protoP2P
```

Comment:
`protocolsByCode[protoP2P.Code] = protoP2P` missing?

Reply:
We set that in `AddProtocol`. Really, the "ipfs" one here isn't necessary. We only need this to set the "p2p" alias.